File: Java/LC20/src/Solution.java
import java.util.Stack;

public class Solution {
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            switch (s.charAt(i)) {
                case '[': case '{': case '(':
                    stack.push(s.charAt(i));
                    break;
                case ']':
                    if (stack.isEmpty() || stack.peek() != '[')
                        return false;
                    stack.pop();
                    break;
                case '}':
                    if (stack.isEmpty() || stack.peek() != '{')
                        return false;
                    stack.pop();
                    break;
                case ')':
                    if (stack.isEmpty() || stack.peek() != '(')
                        return false;
                    stack.pop();
                    break;
            }
        }

        return stack.isEmpty();
    }
}

File: Kotlin/LC20/src/main/kotlin/Solution.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()

        for (i in s.indices) {
            when (s[i]) {
                '(' -> stack.add(s[i])
                ')' -> {
                    if (stack.lastOrNull() == '(') {
                        stack.removeLast()
                    } else {
                        return false
                    }
                }
                '{' -> stack.add(s[i])
                '}' -> {
                    if (stack.lastOrNull() == '{') {
                        stack.removeLast()
                    } else {
                        return false
                    }
                }
                '[' -> stack.add(s[i])
                ']' -> {
                    if (stack.lastOrNull() == '[') {
                        stack.removeLast()
                    } else {
                        return false
                    }
                }
            }
        }

        return stack.isEmpty()
    }
}
